cmd/parallel-manager.go

```diff
@@ -136,7 +150,21 @@ func (p *ParallelManager) monitorProgress() {
 }
 
 func (p *ParallelManager) queueTask(fn func() URLs) {
-	p.queueCh <- fn
+	p.doQueueTask(task{fn: fn})
+}
+
+func (p *ParallelManager) queueTaskWithBarrier(fn func() URLs) {
+	p.doQueueTask(task{fn: fn, barrier: true})
+}
+
+func (p *ParallelManager) doQueueTask(t task) {
+	if t.barrier {
+		p.barrierSync.Wait()
```

Comment:
>  If a WaitGroup is reused to wait for several independent sets of events, new Add calls must happen after all previous Wait calls have returned.

Can you guarantee that?

It seems to me that a goroutine can reach `p.barrierSync.Add(1)` *after* the Wait has been reached here.